File: Scripts/dev/github_comment.py
#!/usr/bin/env python3

import argparse
import json
import subprocess
import sys
import tempfile
from pathlib import Path
from typing import Optional

from github_comment_lib import (
    build_appended_body,
    parse_comment_id,
    parse_github_ref,
    parse_repo_from_remote_url,
    read_body_input,
    resolve_mode,
)


def run_command(args: list[str]) -> str:
    proc = subprocess.run(args, capture_output=True, text=True, check=False)
    if proc.returncode != 0:
        err = proc.stderr.strip() or proc.stdout.strip() or "Unknown command error"
        raise RuntimeError(err)
    return proc.stdout.strip()


def run_gh(args: list[str]) -> str:
    return run_command(["gh", *args])


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Create/append/replace GitHub issue or PR comments with markdown input."
    )
    parser.add_argument("--repo", help="GitHub repo owner/name. Auto-detect when omitted.")
    parser.add_argument("--type", choices=["issue", "pr"], help="Target type.")
    parser.add_argument("--number", type=int, help="Issue/PR number")
    parser.add_argument(
        "--comment",
        help="Comment locator: #issuecomment-123 | issuecomment-123 | full URL | 123",
    )
    parser.add_argument("--mode", choices=["create", "append", "replace"])
    parser.add_argument("--file", help="Markdown file path")
    parser.add_argument("--body", help="Markdown body text")
    parser.add_argument("--ref", help="GitHub issue/pr URL (optionally with #issuecomment-xxx)")
    parser.add_argument("--dry-run", action="store_true", help="Preview resolved operation")
    return parser.parse_args()


def ensure_repo(explicit_repo: Optional[str]) -> str:
    if explicit_repo:
        return explicit_repo

    try:
        from_gh = run_gh(["repo", "view", "--json", "nameWithOwner", "--jq", ".nameWithOwner"])
        if from_gh:
            return from_gh
    except RuntimeError:
        pass

    remote_url = run_command(["git", "config", "--get", "remote.origin.url"])
    return parse_repo_from_remote_url(remote_url)


def write_temp_markdown(content: str) -> Path:
    with tempfile.NamedTemporaryFile("w", encoding="utf-8", suffix=".md", delete=False) as tmp:
        tmp.write(content)
        return Path(tmp.name)


def create_comment(repo: str, number: int, body: str) -> str:
    temp_path = write_temp_markdown(body)
    try:
        return run_gh(
            [
                "api",
                f"repos/{repo}/issues/{number}/comments",
                "-X",
                "POST",
                "-F",
                f"body=@{temp_path}",
                "--jq",
                ".html_url",
            ]
        )
    finally:
        temp_path.unlink(missing_ok=True)


def replace_comment(repo: str, comment_id: str, body: str) -> str:
    temp_path = write_temp_markdown(body)
    try:
        return run_gh(
            [
                "api",
                f"repos/{repo}/issues/comments/{comment_id}",
                "-X",
                "PATCH",
                "-F",
                f"body=@{temp_path}",
                "--jq",
                ".html_url",
            ]
        )
    finally:
        temp_path.unlink(missing_ok=True)


def fetch_comment_body(repo: str, comment_id: str) -> str:
    return run_gh(["api", f"repos/{repo}/issues/comments/{comment_id}", "--jq", ".body"])


def main() -> int:
    args = parse_args()

    parsed_ref = parse_github_ref(args.ref) if args.ref else None

    repo = ensure_repo(args.repo or (parsed_ref["repo"] if parsed_ref else None))
    target_type = args.type or (parsed_ref["type"] if parsed_ref else "issue")

    number = args.number if args.number is not None else (parsed_ref["number"] if parsed_ref else None)
    if not isinstance(number, int) or number <= 0:
        raise ValueError("Issue/PR number is required and must be positive.")

    comment_id: Optional[str] = None
    if args.comment:
        comment_id = parse_comment_id(args.comment)
    elif parsed_ref and parsed_ref.get("comment_id"):
        comment_id = parse_comment_id(str(parsed_ref["comment_id"]))

    mode = resolve_mode(args.mode, comment_id)
    body = read_body_input(args.file, args.body)
    if not body.strip():
        raise ValueError("Input body is empty.")

    if mode in {"append", "replace"} and not comment_id:
        raise ValueError(f"Mode {mode} requires --comment or --ref with #issuecomment-...")

    if args.dry_run:
        preview = {
            "repo": repo,
            "type": target_type,
            "number": number,
            "mode": mode,
            "commentId": comment_id,
            "bodyPreview": body[:160],
        }
        print(json.dumps(preview, ensure_ascii=False, indent=2))
        return 0

    if mode == "create":
        print(create_comment(repo, number, body))
        return 0

    if mode == "replace":
        print(replace_comment(repo, comment_id or "", body))
        return 0

    existing = fetch_comment_body(repo, comment_id or "")
    merged = build_appended_body(existing, body)
    print(replace_comment(repo, comment_id or "", merged))
    return 0


if __name__ == "__main__":
    try:
        raise SystemExit(main())
    except Exception as exc:  # pylint: disable=broad-except
        print(f"[gh-comment] {exc}", file=sys.stderr)
        raise SystemExit(1)

File: Scripts/dev/github-comment.ts
#!/usr/bin/env bun

import { execFileSync } from 'node:child_process';
import { mkdtempSync, rmSync, writeFileSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import {
  buildAppendedBody,
  parseCommentId,
  parseGithubRef,
  parseRepoFromRemoteUrl,
  readBodyInput,
  resolveMode,
  type RefType,
} from './github-comment-lib';

type CliMode = 'create' | 'append' | 'replace';

type Options = {
  repo?: string;
  type?: RefType;
  number?: string;
  comment?: string;
  mode?: CliMode;
  file?: string;
  body?: string;
  ref?: string;
  'dry-run'?: boolean;
  help?: boolean;
};

function fail(message: string): never {
  console.error(`[gh-comment] ${message}`);
  process.exit(1);
}

function runCommand(bin: string, args: string[]): string {
  return execFileSync(bin, args, { encoding: 'utf8' }).trim();
}

function runGh(args: string[]): string {
  return runCommand('gh', args);
}

function printHelp(): void {
  const help = `
Usage:
  bun Scripts/dev/github-comment.ts [options]

Options:
  --repo <owner/repo>             GitHub repo. Optional; auto-detected from origin when omitted.
  --type <issue|pr>               Target type. Default: issue
  --number <id>                   Issue/PR number
  --comment <locator>             Comment locator: #issuecomment-123 | issuecomment-123 | URL | 123
  --mode <create|append|replace>  Operation mode. Default: create (no --comment), append (with --comment)
  --file <path>                   Markdown text file path
  --body <text>                   Markdown text inline
  --ref <github url>              Full issue/pr URL. Can include #issuecomment-xxx
  --dry-run                       Print resolved operation without changing GitHub
  --help                          Show help

Examples:
  bun Scripts/dev/github-comment.ts --type issue --number 93 --file docs/report.md
  bun Scripts/dev/github-comment.ts --type pr --number 89 --comment '#issuecomment-3883010944' --mode replace --file docs/update.md
  bun Scripts/dev/github-comment.ts --ref https://github.com/exomind-team/exomind/issues/93#issuecomment-3883010944 --file docs/add.md --mode append
`;
  console.log(help.trim());
}

function parseArgs(argv: string[]): Options {
  const options: Options = {};

  for (let i = 0; i < argv.length; i += 1) {
    const token = argv[i];
    if (!token.startsWith('--')) {
      throw new Error(`Unexpected positional argument: ${token}`);
    }

    const key = token.slice(2) as keyof Options;
    if (key === 'help' || key === 'dry-run') {
      options[key] = true;
      continue;
    }

    const value = argv[i + 1];
    if (value === undefined || value.startsWith('--')) {
      throw new Error(`Missing value for --${key}`);
    }

    options[key] = value as never;
    i += 1;
  }

  return options;
}

function ensureRepo(explicitRepo: string | undefined): string {
  if (explicitRepo) {
    return explicitRepo;
  }

  try {
    const fromGh = runGh(['repo', 'view', '--json', 'nameWithOwner', '--jq', '.nameWithOwner']);
    if (fromGh) {
      return fromGh;
    }
  } catch {
    // fallback to git remote below
  }

  const remoteUrl = runCommand('git', ['config', '--get', 'remote.origin.url']);
  return parseRepoFromRemoteUrl(remoteUrl);
}

function writeTempBodyFile(content: string): { tempDir: string; bodyPath: string } {
  const tempDir = mkdtempSync(join(tmpdir(), 'gh-comment-'));
  const bodyPath = join(tempDir, 'body.md');
  writeFileSync(bodyPath, content, { encoding: 'utf8' });
  return { tempDir, bodyPath };
}

function createComment(repo: string, number: number, body: string): string {
  const { tempDir, bodyPath } = writeTempBodyFile(body);
  try {
    return runGh([
      'api',
      `repos/${repo}/issues/${number}/comments`,
      '-X',
      'POST',
      '-F',
      `body=@${bodyPath}`,
      '--jq',
      '.html_url',
    ]);
  } finally {
    rmSync(tempDir, { recursive: true, force: true });
  }
}

function replaceComment(repo: string, commentId: string, body: string): string {
  const { tempDir, bodyPath } = writeTempBodyFile(body);
  try {
    return runGh([
      'api',
      `repos/${repo}/issues/comments/${commentId}`,
      '-X',
      'PATCH',
      '-F',
      `body=@${bodyPath}`,
      '--jq',
      '.html_url',
    ]);
  } finally {
    rmSync(tempDir, { recursive: true, force: true });
  }
}

function fetchCommentBody(repo: string, commentId: string): string {
  return runGh(['api', `repos/${repo}/issues/comments/${commentId}`, '--jq', '.body']);
}

function main(): void {
  const options = parseArgs(process.argv.slice(2));
  if (options.help) {
    printHelp();
    return;
  }

  const fromRef = options.ref ? parseGithubRef(options.ref) : undefined;
  const repo = ensureRepo(options.repo ?? fromRef?.repo);
  const type = (options.type ?? fromRef?.type ?? 'issue').toLowerCase() as RefType;
  const number = Number.parseInt(options.number ?? `${fromRef?.number ?? ''}`, 10);

  if (type !== 'issue' && type !== 'pr') {
    throw new Error(`Invalid --type value: ${type}. Expected issue|pr.`);
  }
  if (!Number.isInteger(number) || number <= 0) {
    throw new Error('Issue/PR number is required and must be positive.');
  }

  const commentId = options.comment
    ? parseCommentId(options.comment)
    : (fromRef?.commentId ? parseCommentId(fromRef.commentId) : undefined);

  const mode = resolveMode(options.mode, commentId);
  const incomingBody = readBodyInput(options.file, options.body);

  if ((mode === 'append' || mode === 'replace') && !commentId) {
    throw new Error(`Mode ${mode} requires --comment or --ref with #issuecomment-...`);
  }
  if (!incomingBody.trim()) {
    throw new Error('Input body is empty.');
  }

  if (options['dry-run']) {
    console.log(JSON.stringify({
      repo,
      type,
      number,
      mode,
      commentId: commentId ?? null,
      bodyPreview: incomingBody.slice(0, 160),
    }, null, 2));
    return;
  }

  if (mode === 'create') {
    console.log(createComment(repo, number, incomingBody));
    return;
  }

  if (mode === 'replace') {
    console.log(replaceComment(repo, commentId!, incomingBody));
    return;
  }

  const existing = fetchCommentBody(repo, commentId!);
  const merged = buildAppendedBody(existing, incomingBody);
  console.log(replaceComment(repo, commentId!, merged));
}

try {
  main();
} catch (error) {
  const message = error instanceof Error ? error.message : String(error);
  fail(message);
}
